### DISCUSSION TITLE
Avoid add type `undefined` on data

### DISCUSSION THREAD

**Discussion by Alecell** [TanStack/query#2353] [Ideas]
I'm currently working on a typescript project that gives me this annoying error:
```
No overload matches this call.
  Overload 1 of 2, '(o: {}): string[]', gave the following error.
    Argument of type 'Data<Session & SessionCollection> | {} | undefined' is not assignable to parameter of type '{}'.
      Type 'undefined' is not assignable to type '{}'.
  Overload 2 of 2, '(o: object): string[]', gave the following error.
    Argument of type 'Data<Session & SessionCollection> | {} | undefined' is not assignable to parameter of type 'object'.
      Type 'undefined' is not assignable to type 'object'.ts(2769)
```

I get that this are happening because of that `undefined`:
```ts
Data<Session & SessionCollection> | {} | undefined
```

The point is that my implementation of the service **never** will be `undefined`, what makes me think that this `undefined` came from inside of the typing of the `react-query`. It seems to me that the lib tries to handle the error data of our services for us, but imo that's about to us to type and define. On my case for instance, I have a method that return a key-value object or an empty object:

```ts
static list(
    params: UrlParams,
  ) {
    return campaignRequest(params)
      .collection(EFirestoreCollections.SESSIONS)
      .get()
      .then((res) => res.docs.reduce((session, obj) => {
        const temp = session;

        temp[obj.id] = {
          id: obj.id,
          name: obj.data().name,
        };

        return temp;
      }, {} as ListSessionsState['data']))
      .catch((error) => {
        console.error('Error writing document: ', error);
        return {};
      });
  }
```

If we get the inferred type by the method, without `react-query`: 
```ts
Promise<Data<Session & SessionCollection> | {}> 
```

But for some reason `react-query` infers that it could be `undefined` which isn't true. By far, `react-query` has a really good type inference on the lib so maybe is a good idea to let us to define all the types on the service is the best approach.

But perhaps I'm missing something or doing something wrong, in my situation is there a real case that this `undefined` will truly appears?

**Comment by TkDodo** [comment]
the `data` property returned by `useQuery` can be `undefined` for various reasons:

- if you `disable` the query via `enabled: false`, it will be in `idle` state and data will be undefined
- if you are running a query for the first time, you have no data yet and the query will be in `loading` state with undefined data
- if an error occurs (a rejected Promise, because e.g. your internet connection terminates mid-request), the query will transition to `error` state and if you've never gotten data from the backend, data will be undefined.
